src/algebra/monoid_algebra/degree.lean

```diff
@@ -0,0 +1,150 @@
+  end
+| (f::fs) := begin
+    rw [list.prod_cons, list.map_cons, list.sum_cons],
+    exact (sup_support_mul_le DBm _ _).trans (add_le_add_left (sup_support_list_prod_le _) _)
+  end
+
+lemma le_inf_support_list_prod (DT0 : 0 ≤ DT 0) (DTm : ∀ a b, DT a + DT b ≤ DT (a + b))
+  (l : list (add_monoid_algebra R A)) :
+  (l.map (λ f : add_monoid_algebra R A, f.support.inf DT)).sum ≤ l.prod.support.inf DT :=
+order_dual.of_dual_le_of_dual.mpr $ sup_support_list_prod_le (order_dual.of_dual_le_of_dual.mp DT0)
+  (λ a b, order_dual.of_dual_le_of_dual.mp (DTm _ _)) l
+
+lemma sup_support_pow_le (DB0 : DB 0 ≤ 0) (DBm : ∀ a b, DB (a + b) ≤ DB a + DB b) (n : ℕ)
+  (f : add_monoid_algebra R A) :
+  (f ^ n).support.sup DB ≤ n • (f.support.sup DB) :=
+begin
+  rw [← list.prod_repeat, ←list.sum_repeat],
+  refine (sup_support_list_prod_le DB0 DBm _).trans_eq _,
+  rw list.map_repeat,
+end
+
+lemma le_inf_support_pow (DT0 : 0 ≤ DT 0) (DTm : ∀ a b, DT a + DT b ≤ DT (a + b)) (n : ℕ)
+  (f : add_monoid_algebra R A) :
+  n • (f.support.inf DT) ≤ (f ^ n).support.inf DT :=
+order_dual.of_dual_le_of_dual.mpr $ sup_support_pow_le (order_dual.of_dual_le_of_dual.mp DT0)
+  (λ a b, order_dual.of_dual_le_of_dual.mp (DTm _ _)) n f
+
+end add_monoids
+
+end semiring
+
+section commutative_lemmas
+variables [comm_semiring R] [add_comm_monoid A]
+  [add_comm_monoid B] [covariant_class B B (+) (≤)] [covariant_class B B (function.swap (+)) (≤)]
+  [add_comm_monoid T] [covariant_class T T (+) (≤)] [covariant_class T T (function.swap (+)) (≤)]
+  {DB : A → B} {DT : A → T}
+
+lemma sup_support_multiset_prod_le
+  (DB0 : DB 0 ≤ 0) (DBm : ∀ a b, DB (a + b) ≤ DB a + DB b)
+  (m : multiset (add_monoid_algebra R A)) :
+  m.prod.support.sup DB ≤ (m.map (λ f : add_monoid_algebra R A, f.support.sup DB)).sum :=
+begin
+  induction m using quot.induction_on,
+  rw [multiset.quot_mk_to_coe'', multiset.coe_map, multiset.coe_sum, multiset.coe_prod],
+  exact sup_support_list_prod_le DB0 DBm m,
+end
+
+lemma le_inf_support_multiset_prod
+  (DT0 : 0 ≤ DT 0) (DTm : ∀ a b, DT a + DT b ≤ DT (a + b))
+  (m : multiset (add_monoid_algebra R A)) :
+  (m.map (λ f : add_monoid_algebra R A, f.support.inf DT)).sum ≤ m.prod.support.inf DT :=
+order_dual.of_dual_le_of_dual.mpr $
+  sup_support_multiset_prod_le (order_dual.of_dual_le_of_dual.mp DT0)
+    (λ a b, order_dual.of_dual_le_of_dual.mp (DTm _ _)) m
+
+lemma sup_support_finset_prod_le
+  (DB0 : DB 0 ≤ 0) (DBm : ∀ a b, DB (a + b) ≤ DB a + DB b)
+  (s : finset ι) (f : ι → add_monoid_algebra R A):
+  (∏ i in s, f i).support.sup DB ≤ ∑ i in s, (f i).support.sup DB :=
+(sup_support_multiset_prod_le DB0 DBm _).trans_eq $ congr_arg _ $ multiset.map_map _ _ _
+
+lemma le_inf_support_finset_prod
+  (DT0 : 0 ≤ DT 0) (DTm : ∀ a b, DT a + DT b ≤ DT (a + b))
+  (s : finset ι) (f : ι → add_monoid_algebra R A):
+  ∑ i in s, (f i).support.inf DT ≤ (∏ i in s, f i).support.inf DT :=
+order_dual.of_dual_le_of_dual.mpr $
```

Comment:
This order_dual proof is longer than the direct multiset proof.

Reply:
I used the direct proof, though I could not get it to all work in term-mode.